sam_view.c

```diff
@@ -560,211 +560,428 @@ static const char *copied_tags[] = { "RG", "BC", "QT", NULL };
+            print_error_errno("Cannot write to r0 file \"%s\"", opts->fnr1);
+            free(state);
+            return true;
+        }
     }
-    b = bam_init1();
-    if (b == NULL) {
-        perror(NULL);
-        bam_hdr_destroy(h);
-        return 1;
+
+    if (opts->fnr1) {
+        state->fpr[1] = fopen(opts->fnr1, "w");
+        if (state->fpr[1] == NULL) {
+            print_error_errno("Cannot write to r1 file \"%s\"", opts->fnr1);
+            free(state);
+            return true;
+        }
     }
-    buf = NULL;
-    max_buf = 0;
 
-    fpse = NULL;
-    if (fnse) {
-        fpse = fopen(fnse,"w");
-        if (fpse == NULL) {
-            print_error_errno("Cannot write to singleton file \"%s\"", fnse);
-            bam_hdr_destroy(h);
-            return 1;
+    if (opts->fnr2) {
+        state->fpr[2] = fopen(opts->fnr2, "w");
+        if (state->fpr[2] == NULL) {
+            print_error_errno("Cannot write to r2 file \"%s\"", opts->fnr2);
+            free(state);
+            return true;
         }
     }
 
-    int64_t n_singletons = 0, n_reads = 0;
-    char* previous = NULL;
-    kstring_t linebuf = { 0, 0, NULL };
-    kputsn("", 0, &linebuf);
+    state->h = sam_hdr_read(state->fp);
+    if (state->h == NULL) {
+        fprintf(stderr, "Failed to read header for \"%s\"\n", opts->fn_input);
+        free(state);
+        return true;
+    }
 
-    while (sam_read1(fp, h, b) >= 0) {
-        if (b->core.flag&(BAM_FSECONDARY|BAM_FSUPPLEMENTARY)) continue; // skip secondary and supplementary alignments
-        ++n_reads;
+    *state_out = state;
+    return false;
+}
 
-        int i;
-        int32_t qlen = b->core.l_qseq;
-        assert(qlen >= 0);
-        uint8_t* seq;
-        uint8_t* qual = bam_get_qual(b);
-        const uint8_t *oq = NULL;
-        if (use_oq) oq = bam_aux_get(b, "OQ");
-        bool has_qual = (qual[0] != 0xff || (use_oq && oq)); // test if there is quality
-
-        // If there was a previous readname
-        if ( fpse && previous ) {
-            if (!strcmp(bam_get_qname(b), previous ) ) {
-                fputs(linebuf.s, stdout); // Write previous read
-                free(previous);
-                previous = NULL;
-            } else { // Doesn't match it's a singleton
-                ++n_singletons;
-                fputs(linebuf.s, fpse);  // Write previous read to singletons
-                free(previous);
-                previous = strdup(bam_get_qname(b));
-            }
-        } else {
-            fputs(linebuf.s, stdout); // Write pending read
-            if (fpse) previous = strdup(bam_get_qname(b));
-        }
+static bool destroy_state(const bam2fq_opts_t *opts, bam2fq_state_t *state, int* status)
+{
+    bam_hdr_destroy(state->h);
+    check_sam_close(state->fp, opts->fn_input, "file", status);
+    if (state->fpse && fclose(state->fpse)) { print_error_errno("Error closing singleton file \"%s\"", opts->fnse); return true; }
```

Comment:
Don't return here, set an error flag instead and return it at the end.  That way everything will be closed and freed, even if you get an error.